azurelinuxagent/ga/collect_logs.py

```diff
@@ -0,0 +1,225 @@
+# Microsoft Azure Linux Agent
+#
+# Copyright 2020 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+
+import datetime
+import os
+import sys
+import threading
+import time
+
+import azurelinuxagent.common.conf as conf
+from azurelinuxagent.common import logger
+from azurelinuxagent.common.event import elapsed_milliseconds, add_event, WALAEventOperation
+from azurelinuxagent.common.future import ustr
+from azurelinuxagent.common.logcollector import COMPRESSED_ARCHIVE_PATH
+from azurelinuxagent.common.protocol.util import get_protocol_util
+from azurelinuxagent.common.utils import shellutil
+from azurelinuxagent.common.utils.shellutil import get_python_cmd
+from azurelinuxagent.common.version import PY_VERSION_MAJOR, PY_VERSION_MINOR, AGENT_NAME, CURRENT_VERSION
+from azurelinuxagent.ga.periodic_operation import PeriodicOperation
+
+
+def get_collect_logs_handler():
+    return CollectLogsHandler()
+
+
+class CollectLogsHandler(object):
+    """
+    Periodically collects and uploads logs from the VM to the host.
+    """
+
+    _THREAD_NAME = "CollectLogsHandler"
+
+    @staticmethod
+    def get_thread_name():
+        return CollectLogsHandler._THREAD_NAME
+
+    def __init__(self):
+        self.protocol = None
+        self.protocol_util = None
+        self.event_thread = None
+        self.should_run = True
+        self.last_state = None
+
+        self._periodic_operations = [
+            PeriodicOperation("collect_and_send_logs", self.collect_and_send_logs, conf.get_collect_logs_period())
+        ]
+
+    def _log_collection_allowed(self):
+        # There are three conditions that need to be met in order to allow periodic log collection:
+        # 1) It should be enabled in the configuration.
+        # 2) The system must be using systemd to manage services. Needed for resource limiting of the log collection.
+        # 3) The python version must be greater than 2.6 in order to support the ZipFile library used when collecting.
+        conf_enabled = conf.get_collect_logs()
+        systemd_present = os.path.exists("/run/systemd/system/")
+        supported_python = PY_VERSION_MINOR >= 7 if PY_VERSION_MAJOR == 2 else PY_VERSION_MAJOR == 3
+        is_allowed = conf_enabled and systemd_present and supported_python
+
+        if self.last_state != is_allowed:
```

Comment:
Is there a chance that any of these requirements can be changed without restarting the agent? If not then we should remove the `self.last_state` and just log this once (maybe before adding the thread handler in the `update.py` file)

Reply:
That is a good point, and yes, the agent needs to be restarted in order for any of the changes to the three conditions to take effect. Will log it once in the `update.py` module. Thanks!